// @virtuals-protocol/game-imagegen-plugin/dist/example.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const dotenv_1 = require("dotenv");
(0, dotenv_1.config)({ path: './.env' });
const game_1 = require("@virtuals-protocol/game");
const _1 = __importDefault(require("."));
const imageGenPlugin = new _1.default({
    apiClientConfig: {
        apiKey: process.env.TOGETHER_API_KEY, // Default key: UP-17f415babba7482cb4b446a1
    },
});
// Create an agent with the worker
const agent = new game_1.GameAgent((_a = process.env.GAME_API_KEY) !== null && _a !== void 0 ? _a : "", {
    name: "Image Generation Worker",
    goal: "Generate an anime-style character image with Twitter logo.",
    description: "You are an AI agent specialized in generating images. You can create images based on text prompts using Together AI's FLUX schnell model.",
    workers: [
        imageGenPlugin.getWorker({}),
    ],
});
(async () => {
    agent.setLogger((agent, message) => {
        console.log(`-----[${agent.name}]-----`);
        console.log(message);
        console.log("\n");
    });
    await agent.init();
    while (true) {
        await agent.step({
            verbose: true,
        });
    }
})();
